sljit_src/sljitNativeS390X.c

```diff
@@ -0,0 +1,2817 @@
+			FAIL_IF(load_word(compiler, src2_r, src2, src2w, tmp1, op & SLJIT_I32_OP));
+		}
+		/* TODO(mundaym): distinct operand facility where needed */
+		if (src1_r != dst_r && src1_r != tmp0) {
+			FAIL_IF(push_inst(compiler, (op & SLJIT_I32_OP) ?
+				lr(tmp0, src1_r) : lgr(tmp0, src1_r)));
+			src1_r = tmp0;
+		}
+		sljit_ins ins = 0;
+		switch (GET_OPCODE(op) | (op & SLJIT_I32_OP)) {
+		/* 64-bit ops */
+		case SLJIT_ADD: ins = signed_flags ?
+			agr(src1_r, src2_r) :
+			algr(src1_r, src2_r);
+			break;
+		case SLJIT_SUB: ins = signed_flags ?
+			sgr(src1_r, src2_r) :
+			slgr(src1_r, src2_r);
+			break;
+		case SLJIT_ADDC:
+			ins = alcgr(src1_r, src2_r);
+			break;
+		case SLJIT_SUBC:
+			ins = slbgr(src1_r, src2_r);
+			break;
+		case SLJIT_MUL:
+			ins = msgr(src1_r, src2_r);
+			break;
+		case SLJIT_AND:
+			ins = ngr(src1_r, src2_r);
+			break;
+		case SLJIT_OR:
+			ins = ogr(src1_r, src2_r);
+			break;
+		case SLJIT_XOR:
+			ins = xgr(src1_r, src2_r);
+			break;
+		/* 32-bit ops */
+		case SLJIT_ADD32:
+			ins = signed_flags ?
+				ar(src1_r, src2_r):
+				alr(src1_r, src2_r);
+			break;
+		case SLJIT_SUB32:
+			ins = signed_flags ?
+				sr(src1_r, src2_r):
+				slr(src1_r, src2_r);
+			break;
+		case SLJIT_ADDC32:
+			ins = alcr(src1_r, src2_r);
+			break;
+		case SLJIT_SUBC32:
+			ins = slbr(src1_r, src2_r);
+			break;
+		case SLJIT_MUL32:
+			ins = msr(src1_r, src2_r);
+			break;
+		case SLJIT_AND32:
+			ins = nr(src1_r, src2_r);
+			break;
+		case SLJIT_OR32:
+			ins = or(src1_r, src2_r);
+			break;
+		case SLJIT_XOR32:
+			ins = xr(src1_r, src2_r);
+			break;
+		default:
+			SLJIT_UNREACHABLE();
+		}
+		FAIL_IF(push_inst(compiler, ins));
+		if (src1_r != dst_r)
+			FAIL_IF(push_inst(compiler, (op & SLJIT_I32_OP) ?
+				lr(dst_r, src1_r) : lgr(dst_r, src1_r)));
+
+	}
+
+	/* write condition code to emulated flag register */
+	if (op & VARIABLE_FLAG_MASK)
+		FAIL_IF(push_inst(compiler, ipm(flag_r)));
+
+	/* write zero flag to emulated flag register */
+	if (op & SLJIT_SET_Z)
+		FAIL_IF(push_store_zero_flag(compiler, op, dst_r));
+
+	/* finally write the result to memory if required */
+	if (dst & SLJIT_MEM) {
+		SLJIT_ASSERT(dst_r != tmp1);
+		/* TODO(carenas): s/FAIL_IF/ return */
+		FAIL_IF(store_word(compiler, dst_r, dst, dstw, tmp1, op & SLJIT_I32_OP));
+	}
+
+	return SLJIT_SUCCESS;
+}
+
+SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_op_src(
+	struct sljit_compiler *compiler,
+	sljit_s32 op, sljit_s32 src, sljit_sw srcw)
+{
+	sljit_gpr src_r;
+
+	CHECK_ERROR();
+	CHECK(check_sljit_emit_op_src(compiler, op, src, srcw));
+	ADJUST_LOCAL_OFFSET(src, srcw);
+
+	switch (op) {
+	case SLJIT_FAST_RETURN:
+		src_r = FAST_IS_REG(src) ? gpr(src) : tmp1;
+		if (src & SLJIT_MEM)
+			FAIL_IF(load_word(compiler, tmp1, src, srcw, tmp1, 0));
+
+		return push_inst(compiler, br(src_r));
+	case SLJIT_SKIP_FRAMES_BEFORE_FAST_RETURN:
+	case SLJIT_PREFETCH_L1:
+	case SLJIT_PREFETCH_L2:
+	case SLJIT_PREFETCH_L3:
+	case SLJIT_PREFETCH_ONCE:
```

Comment:
Are there no prefect instructions in s390x or they have just been not implemented yet? If the latter, please add a todo comment.